src/core2cuda.rkt

```diff
@@ -0,0 +1,152 @@
+#lang racket
+
+(require generic-flonum)
+(require "imperative.rkt")
+
+(provide cuda-header
+         cuda-supported
+         cuda-reserved
+         core->cuda
+         type->cuda
+         params->cuda      ; required by core2h.rkt
+         set-cuda-header!
+         set-unknown->cuda!)
+
+(define cuda-header (const "#include <fenv.h>\n#include <math.h>\n#include <stdint.h>\n__device__ double e = 2.71828182845904523536;\n\n"))
+(define cuda-supported
+  (supported-list
+    (invert-op-proc (curry set-member? '(array dim size ref for for* tensor tensor*)))
+    fpcore-consts
+    (curry set-member? '(binary32 binary64 binary80 integer))
+    (invert-rnd-mode-proc (curry equal? 'nearestAway))
+    #f))
+
+(define (default-unknown->cuda ctx op args)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (format "~a~a(~a)"
+          op
+          (cuda-type->suffix type)
+          (string-join args ", ")))
+
+(define unknown->cuda (make-parameter default-unknown->cuda))
+
+(define (set-cuda-header! gen-proc)
+  (set! cuda-header (gen-proc cuda-header)))
+
+(define (set-unknown->cuda! gen-proc)
+  (unknown->cuda (gen-proc (unknown->cuda))))
+
+(define cuda-reserved  ; Language-specific reserved names (avoid name collisions)
+  '(auto break case char const continue default
+    do double else enum extern float for goto if
+    inline int long register restrict return
+    short signed sizeof static struct switch
+    typedef union unsigned void volatile while))
+
+(define/match (cuda-type->suffix type)
+  [("int64_t") ""]
+  [("double") ""]
+  [("float") "f"]
+  [("long double") "l"])
+
+(define/match (type->cuda type)
+  [('binary64) "double"]
+  [('binary32) "float"]
+  [('binary80) "long double"]
+  [('boolean) "int"]
+  [('integer) "int64_t"])
+
+(define (operator->cuda op args ctx)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (match op
+   [(or 'isinf 'isnan 'isfinite 'isnormal 'signbit)
+    (format "~a~a" op args)]
+   [_
+    ((unknown->cuda) ctx op args)]))
+  
+(define (binary80->string x)
+  (parameterize ([gfl-exponent 15] [gfl-bits 80])
+    (let ([s (gfl->string (gfl x))])
+      (if (string-contains? s ".") s (string-append s ".0")))))
+
+(define (constant->cuda x ctx)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (match x
+   [(or 'TRUE 'FALSE (? hex?)) (~a x)]
+   [(or 'M_1_PI 'M_2_PI 'M_2_SQRTPI 'INFINITY 'NAN)
```

Comment:
Do these constants actually exist? I'm not really sure, I see some CUDA stuff online talk about a different header you're supposed to get these from.

Reply:
Ok so from looking this up too, `math_constants.h` defines constants that can be used. We could use 

CUDART_INF from math_constants.h and CUDART_NAN from math_constants.h for INFINITY and NAN

M_1_PI and M_2_PI: aren't not directly provided by CUDA, but we can define them after including the CUDA runtime -> `<cuda_runtime.h>` then by doing:

#define M_1_PI (1.0 / CUDART_PI)
#define M_2_PI (2.0 / CUDART_PI)
#define M_2_SQRTPI (2.0 / sqrt(CUDART_PI))

should I make the changes below to the (define (constant->cuda x ctx) expression and include the necessary header files at the top? Or should I just define those in the header and then replace their names on line 74?

```
(define (constant->cuda x ctx)
  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
  (match x
    [(or 'TRUE 'FALSE (? hex?)) (~a x)]
    ; Special case the CUDA-specific infinity and nan constants
    ['INFINITY (format "((~a) CUDART_INF)" type)]
    ['NAN (format "((~a) CUDART_NAN)" type)]
    ; Handle the derived PI constants using CUDART_PI
    ['M_1_PI (format "((~a) (1.0 / CUDART_PI))" type)]
    ['M_2_PI (format "((~a) (2.0 / CUDART_PI))" type)]
    ['M_2_SQRTPI (format "((~a) (2.0 / sqrt(CUDART_PI)))" type)]
    [(? number?)
     (match type
       ["int64_t" (~a (inexact->exact x))]
       [_ (format "~a~a" (real->double-flonum x) (cuda-type->suffix type))])]
    [(? symbol?) (format "((~a) M_~a)" type x)]))
```